fix(compiler-cli): check split two way binding

danieltre23 opened the pull request:
Check for split two way binding when output is not declared to make error message clearer.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: 39300


## What is the new behavior?
The new error message for a banana-in-a-box binding with the output not declared should be "The output {outputName} for the two way binding is not declared."

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

dgp1130 commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -1031,7 +1031,7 @@ export class TcbDirectiveOutputsOp extends TcbOp {
   constructor(
       private tcb: Context, private scope: Scope, private element: TmplAstElement,
-      private claimedOutputs: Set<string>) {
+      private claimedOutputs: Set<string>, private claimedInputs: Set<string>) {
```
Nit: I would swap the `claimedInputs` and `claimedOutputs` parameters so inputs comes first. Otherwise I think it would be pretty easy for users to accidentally pass their arguments backwards.

dgp1130 commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -1049,6 +1049,17 @@ class TcbUnclaimedOutputsOp extends TcbOp {
       }
 
+      // Check for split two way binding with unclaimed outputs
+      let inputName = output.name.replace('Change', '');
```
We can probably limit this to just outputs which *end* with `Change`, otherwise a name like `ChangeDataEvent` would become `DataEvent` and lead to strange behavior.

I think we can use a regex to only replace a `Change` which is the last part of the string:

```typescript
output.name.replace(/Change$/, '');
```

dgp1130 commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -1049,6 +1049,17 @@ class TcbUnclaimedOutputsOp extends TcbOp {
         continue;
       }
 
+      // Check for split two way binding with unclaimed outputs
```
Instead of removing `Change` and then looking for an input with that name, we should look for an input *only if* the output ends with `Change`. Otherwise something like:

```typescript
class MyComponent {
  @Input() @Output() something: string;
}
```

Would lead to strange behavior. Admittedly that's probably its own bug, but I imagine we'd want to handle that separately from this.

dgp1130 commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -1049,6 +1049,17 @@ class TcbUnclaimedOutputsOp extends TcbOp {
+      let inputName = output.name.replace('Change', '');
+      if (this.claimedInputs.has(inputName)) {
+        let input = this.element.inputs.find(input => input.name == inputName)
+        if (input?.sourceSpan == output.sourceSpan) {
```
Nit: Since already know the `inputName` is claimed, if we don't get an `input`, that's an internal compiler error. I'm thinking we should either assert that `input` is defined (not sure of the best way to do that in the compiler) or use `input!.sourceSpan` so we get a real error.

dgp1130 commented on packages/compiler-cli/test/ngtsc/template_typecheck_spec.ts:
```diff
@@ -338,6 +338,42 @@ export declare class AnimationEvent {
+
+        export class ChildCmp {
+          @Input() value = 0;
+          @Output() incrementValue = new EventEmitter<number>();
```
Nit: Consider adding a comment here which points out that this name should be `valueChange`, but the user got it wrong which causes the error. This will make it more clear exactly what mistake this is testing for.

dgp1130 commented on packages/compiler-cli/src/ngtsc/typecheck/src/oob.ts:
```diff
@@ -211,6 +220,15 @@ export class OutOfBandDiagnosticRecorderImpl implements OutOfBandDiagnosticRecor
+
+  splitTwoWayBinding(templateId: TemplateId, output: TmplAstBoundEvent): void {
+    const mapping = this.resolver.getSourceMapping(templateId);
+    const errorMsg = `The output ${output.name} for the two way binding is not declared.`;
```
Nit: I'm thinking we should include a link to https://angular.io/guide/two-way-binding#how-two-way-binding-works which gives a much more complete explanation of the two-way binding contract. That would help users understand exactly what they did wrong and how to fix it.

dgp1130 reviewed (commented):
Great job Daniel! This looks really cool. Just a few nitpicks and a couple edge cases to consider and this should be good to go.

One general nit: Prefer `const` over `let` wherever possible, just to reduce the possibility of mutation where its not needed.

dgp1130 commented:
All the code looks good to me, and I kicked off a google3 presubmit, but it looks like it has a few legitimate failures: http://test/OCL:380647145:BASE:380647172:1624306026646:e07d2f4f.

Looks like there's a lot of code which two-way binds to `model` properties that don't have a `modelChange` event. I'm surprised this wasn't an error before, do we special case the `model` property in some way? I'm wondering if we should exempt it from this check or else we'll likely need to roll this out behind a flag. Thoughts, @alxhub?

danieltre23 commented:
> All the code looks good to me, and I kicked off a google3 presubmit, but it looks like it has a few legitimate failures: http://test/OCL:380647145:BASE:380647172:1624306026646:e07d2f4f.
> 
> Looks like there's a lot of code which two-way binds to `model` properties that don't have a `modelChange` event. I'm surprised this wasn't an error before, do we special case the `model` property in some way? I'm wondering if we should exempt it from this check or else we'll likely need to roll this out behind a flag. Thoughts, @alxhub?

I think the error should only show when the strictTemplates flag is true, I noticed that the error in the issue #39300 is only for strictTemplates, so I'll add that check.

danieltre23 committed:
fix(compiler-cli): check split two way binding

Check for split two way binding when output is not declared to make error message clearer.

dgp1130 commented:
Passing presubmit: http://test/OCL:380647145:BASE:380826857:1624379963990:c3d8b02d

dgp1130 commented:
Ran a global presubmit since strict templates is not widely used internally and I wanted to make sure we were checking at least some strict templates targets, everything looks good: http://test/OCL:380970929:BASE:380971055:1624432874430:81de2e8d.

dgp1130 reviewed (approved):
Forgot to approve after the google3 tests passed.

alxhub commented on packages/compiler-cli/src/ngtsc/diagnostics/src/error_code.ts:
```diff
@@ -166,6 +166,11 @@ export enum ErrorCode {
   DUPLICATE_VARIABLE_DECLARATION = 8006,
 
+  /**
+   * A template has a double binding in which the output is not declared.
```
I think we need to elaborate more on what this means, and clarify:

* We're talking about a two way binding, where a single syntactical element creates two bindings (one to an input, one to an output).
* This binding is "split", meaning that the input and output are interpreted as going to different places (as opposed to the same component or directive).

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/oob.ts:
```diff
@@ -211,6 +220,17 @@ export class OutOfBandDiagnosticRecorderImpl implements OutOfBandDiagnosticRecor
         ngErrorCode(ErrorCode.SUGGEST_SUBOPTIMAL_TYPE_INFERENCE), message));
   }
+
+  splitTwoWayBinding(templateId: TemplateId, output: TmplAstBoundEvent): void {
```
I think we're missing an important distinction here.

The fundamental issue is that for a true two way binding, both the input and output halves should go to the same place (either a directive of some kind, or to DOM). If we see a usage where this _isn't_ the case, it doesn't automatically mean that the component is missing an `@Output`. It might very well be that the user has misunderstood the component's API, or that the `@Input` is actually the part that's in the wrong (maybe the output goes to a directive, but the input doesn't?).

Right now, we're only really catching the case where the `@Input` goes to a directive, but the `@Output` goes to DOM. In this case, I think we should give an error that:

* Reports the problem, e.g. "The input and output halves of this two-way binding don't bind to the same target"
* Describes where each of the bindings goes. For example, if the input half of the binding goes to a component/directive, we can add an additional pointer to the diagnostic saying "the input half of the binding goes to this directive".

Our error shouldn't favor any particular interpretation (e.g. that an `@Output` is missing somewhere) but should instead describe the actual situation. In the error guide, we can then make suggestions about what _might_ be wrong in the program to have caused the error.

atscott replied:
Hi @danieltre23, unresolving this again. Please add the following test to your suite, which should pass when this comment is resolved: 

```
      it('when input and output go to different directives', () => {
        env.tsconfig({strictTemplates: true});
        env.write('test.ts', `
        import {Component, Input, NgModule, Output, Directive} from '@angular/core';

        @Component({
          selector: 'test',
          template: '<child-cmp [(value)]="counterValue"></child-cmp>',
        })
        export class TestCmp {
          counterValue = 0;
        }

        @Directive({
          selector: 'child-cmp'
        })
        export class ChildCmpDir {
          @Output() valueChange: any;
        }

        @Component({
          selector: 'child-cmp',
          template: '',
        })
        export class ChildCmp {
          @Input() value = 0;
        }

        @NgModule({
          declarations: [TestCmp, ChildCmp, ChildCmpDir],
        })
        export class Module {}
      `);
        const diags = env.driveDiagnostics();
        expect(diags.length).toBe(1);
        expect(diags[0].code).toBe(ngErrorCode(ErrorCode.SPLIT_TWO_WAY_BINDING));
        expect(getSourceCodeForDiagnostic(diags[0])).toBe('[(value)]="counterValue"');
        expect(diags[0].relatedInformation!.length).toBe(2);
      });
```

alxhub commented on packages/compiler-cli/test/ngtsc/template_typecheck_spec.ts:
```diff
@@ -338,6 +338,44 @@ export declare class AnimationEvent {
+        export class ChildCmp {
+          @Input() value = 0;
+          // Output name should be valueChange
+          @Output() incrementValue = new EventEmitter<number>();
```
We shouldn't actually need an `@Output` here at all - the test is equally valid and effective if this is missing.

alxhub reviewed (changes requested):
Overall I think the mechanics of this PR are solid. Let's tweak the error message a bit to more accurately reflect the situation in the template, and then this should be good to go.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/oob.ts:
```diff
@@ -211,6 +223,21 @@ export class OutOfBandDiagnosticRecorderImpl implements OutOfBandDiagnosticRecor
+      templateId: TemplateId, input: BoundAttribute, inputProptertyName: string|null,
+      outputProptertyName: string|null): void {
+    const mapping = this.resolver.getSourceMapping(templateId);
+    const errorMsg = `The input and output of the two-way binding '${
```
This error is getting better! Let's adjust a few more things:

1. Inputs and outputs are features of components and directives. In the context of a binding, you bind to a _property_ or to an _event_. Inputs are a feature of components that can receive property bindings, outputs are a feature of components that can raise events. So using the property/event terminology is more accurate in this context.

2. It's not a great developer experience for us to say "the output is bound to null". In this case, we should tell them that the property half of the binding targets a component, whereas the event binding is targeting the DOM.

3. The property/event names for the binding are wholly predictable, as they're determined by the binding the user wrote. If the user wrote `[(ngModel)]` then the property name will be `ngModel` and the event name will be `ngModelChange`. It'd be more useful to name the component(s) which are receiving the binding(s).

. `ts.Diagnostic` has a way of attaching related information beyond the primary error message. This is extremely useful for building a good developer experience for an error, as we can include pointers to other locations in the code that are relevant to the error. For example, in this case we could point to the component that's accepting the property binding. Some of the other OOB diagnostics make use of this facility - you can use them as examples.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -1049,6 +1049,32 @@ class TcbUnclaimedOutputsOp extends TcbOp {
+          // Get input bindings from template AST. If not found, then the user
+          // did not bind an input, so this is not a two-way data binding.
+          const input = this.element.inputs.find(input => input.name == inputName);
+          if (input && input.sourceSpan == output.sourceSpan) {
```
In the compiler we prefer to check values explicitly rather than relying on truthiness. Additionally, in JavaScript you should always prefer `===` to `==` as the latter has some really unintuitive behavior in specific cases. For example, `3 == '3'` is true.


```suggestion
          if (input !== undefined && input.sourceSpan === output.sourceSpan) {
```

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -1049,6 +1049,32 @@ class TcbUnclaimedOutputsOp extends TcbOp {
+        if (this.claimedInputs.has(inputName)) {
+          // Get input bindings from template AST. If not found, then the user
+          // did not bind an input, so this is not a two-way data binding.
+          const input = this.element.inputs.find(input => input.name == inputName);
```
(same as below, prefer `===`)

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -1049,6 +1049,32 @@ class TcbUnclaimedOutputsOp extends TcbOp {
+          const input = this.element.inputs.find(input => input.name == inputName);
+          if (input && input.sourceSpan == output.sourceSpan) {
+            const inputConsumer = this.tcb.boundTarget.getConsumerOfBinding(input);
+            let inputProptertyName = null;
```
```suggestion
            let inputPropertyName = null;
```
and throughout

atscott commented on packages/compiler-cli/src/ngtsc/typecheck/src/oob.ts:
```diff
@@ -211,6 +223,21 @@ export class OutOfBandDiagnosticRecorderImpl implements OutOfBandDiagnosticRecor
+      outputProptertyName: string|null): void {
+    const mapping = this.resolver.getSourceMapping(templateId);
+    const errorMsg = `The input and output of the two-way binding '${
+        input.name}' are not binded to the same target.
```
Nit: these should be “bound” not “binded”

atscott commented on packages/compiler-cli/test/ngtsc/template_typecheck_spec.ts:
```diff
@@ -338,6 +338,44 @@ export declare class AnimationEvent {
+      const diags = env.driveDiagnostics();
+      expect(diags.length).toBe(1);
+      expect(diags[0].messageText)
+          .toEqual(
```
We should probably assert the error code here rather than the text of the message. We shouldn’t have to keep updating this test assertion if we revise the wording of the error. 

Edit: note that the error codes are a relatively new thing we introduced so the old tests probably didn’t have that option and May no longer be the best example of what the test should assert.

atscott commented on packages/compiler-cli/src/ngtsc/diagnostics/src/error_code.ts:
```diff
@@ -166,6 +166,14 @@ export enum ErrorCode {
+   * A template has a two way binding (two bindings created by a single syntactial element)
+   * in which the input and output are going to different places.
+   *
+   * <div [(ngModel)]="..."/>
```
I think this example should either be removed completely or changed. `ngModel` typically is not a split two way binding.

atscott reviewed (approved):
Adding my approval. Left a couple comments that could be addressed but they're minor/trivial.

danieltre23 commented:
Flaky failures in presubmit: http://test/OCL:382577037:BASE:382578882:1625167383976:5adc77a7 forced green.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/oob.ts:
```diff
@@ -211,6 +219,55 @@ export class OutOfBandDiagnosticRecorderImpl implements OutOfBandDiagnosticRecor
+
+  splitTwoWayBinding(
+      templateId: TemplateId, input: BoundAttribute, output: TmplAstBoundEvent,
+      inputConsumerIdentifier: ts.Identifier, outputConsumer: ts.Identifier|TmplAstElement): void {
```
I would suggest passing the `ClassDeclaration` here instead of the `ts.Identifer`, as logically the binding goes to the class. Let `splitToWayBinding` decide how it wants to identify the component.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/oob.ts:
```diff
@@ -211,6 +219,55 @@ export class OutOfBandDiagnosticRecorderImpl implements OutOfBandDiagnosticRecor
+      templateId: TemplateId, input: BoundAttribute, output: TmplAstBoundEvent,
+      inputConsumerIdentifier: ts.Identifier, outputConsumer: ts.Identifier|TmplAstElement): void {
+    const mapping = this.resolver.getSourceMapping(templateId);
+    const errorMsg = `The property and event of the two-way binding '${
```
Let's refer to them as the "property and event halves of the two-way binding"

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/oob.ts:
```diff
@@ -211,6 +219,55 @@ export class OutOfBandDiagnosticRecorderImpl implements OutOfBandDiagnosticRecor
+      });
+    }
+
+    diagnostic.relatedInformation = relatedDiagnostics;
```
We shouldn't need to modify `diagnostic.relatedInformation` here. `makeTemplateDiagnostic` should be able to fully construct the diagnostic with all information already present. If it can't, we can modify it to be more flexible in that regard.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/oob.ts:
```diff
@@ -211,6 +219,55 @@ export class OutOfBandDiagnosticRecorderImpl implements OutOfBandDiagnosticRecor
+        messageText: `The output binding is to a native event called '${input.name}' on the '${
+            outputConsumer.name}' DOM element.
+      
+      Are you missing an output declaration called '${output.name}'?`,
```
We should only show this hint if the component belongs to the user, and doesn't come from a library.

`ts.SourceFile` has `.isDeclarationFile`, which signifies that the source file is a `.d.ts` file from a library instead of a `.ts` file that the user wrote, which can distinguish these two cases.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/oob.ts:
```diff
@@ -211,6 +219,55 @@ export class OutOfBandDiagnosticRecorderImpl implements OutOfBandDiagnosticRecor
+    let relatedDiagnostics = [{
+      category: ts.DiagnosticCategory.Message,
+      code: 0,
+      file: mapping.node.getSourceFile(),
```
I think this is wrong - `inputConsumerIdentifier` may be in an entirely different file than the template itself.

danieltre23 commented:
Flaky failures in presubmit: http://test/OCL:382577037:BASE:382802023:1625257425653:73bd3481 forced green.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/oob.ts:
```diff
@@ -211,6 +223,50 @@ export class OutOfBandDiagnosticRecorderImpl implements OutOfBandDiagnosticRecor
+    let relatedMessages = [];
+
+    relatedMessages.push({
+      text: `The property is bound to the '${inputConsumer.name.getText()}' component.`,
```
```suggestion
      text: `The property half of the binding is to the '${inputConsumer.name.getText()}' component.`,
```

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/oob.ts:
```diff
@@ -211,6 +223,50 @@ export class OutOfBandDiagnosticRecorderImpl implements OutOfBandDiagnosticRecor
+    });
+
+    if (outputConsumer instanceof TmplAstElement) {
+      let message = `The output binding is to a native event called '${input.name}' on the '${
```
```suggestion
      let message = `The event half of the binding is to a native event called '${input.name}' on the '${
```

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/oob.ts:
```diff
@@ -211,6 +223,50 @@ export class OutOfBandDiagnosticRecorderImpl implements OutOfBandDiagnosticRecor
+
+    if (outputConsumer instanceof TmplAstElement) {
+      let message = `The output binding is to a native event called '${input.name}' on the '${
+          outputConsumer.name}' DOM element.`;
```
Consider maybe `<${outputConsumer.name}>` instead of single quotes?

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/oob.ts:
```diff
@@ -211,6 +223,50 @@ export class OutOfBandDiagnosticRecorderImpl implements OutOfBandDiagnosticRecor
+      }
+      relatedMessages.push({
+        text: message,
+        start: outputConsumer.sourceSpan.start.offset,
```
Let's be a little tricky here, and only highlight the tag name. We can do that by computing the start and end positions of the tag name.

Since the element looks like `<tag-name ....>`, we see that the tag name starts at the start span + 1, and ends at that start + the length of the tag name. That should be enough to find the tag name span.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/oob.ts:
```diff
@@ -211,6 +223,50 @@ export class OutOfBandDiagnosticRecorderImpl implements OutOfBandDiagnosticRecor
+      });
+    } else {
+      relatedMessages.push({
+        text: `The event is bound to the '${outputConsumer.name.getText()}' component.`,
```
```suggestion
        text: `The event is bound to the '${outputConsumer.name.text}' component.`,
```

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/oob.ts:
```diff
@@ -211,6 +223,50 @@ export class OutOfBandDiagnosticRecorderImpl implements OutOfBandDiagnosticRecor
+        input.name}' are not bound to the same target.
+            Find more at https://angular.io/guide/two-way-binding#how-two-way-binding-works`;
+
+    let relatedMessages = [];
```
It's always good to explicitly type arrays like this. Also, this can be `const` since we don't need to change its value.

alxhub commented on packages/compiler-cli/test/ngtsc/template_typecheck_spec.ts:
```diff
@@ -338,6 +338,81 @@ export declare class AnimationEvent {
+      const diags = env.driveDiagnostics();
+      expect(diags.length).toBe(1);
+      expect(diags[0].code).toBe(ngErrorCode(ErrorCode.SPLIT_TWO_WAY_BINDING));
+      expect(getSourceCodeForDiagnostic(diags[0])).toBe('[(value)]="counterValue"');
```
We should assert that the related messages are also pointing to the right things.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/oob.ts:
```diff
@@ -211,6 +223,50 @@ export class OutOfBandDiagnosticRecorderImpl implements OutOfBandDiagnosticRecor
+
+
+    this._diagnostics.push(makeTemplateDiagnostic(
+        templateId, mapping, input.sourceSpan, ts.DiagnosticCategory.Error,
```
```suggestion
        templateId, mapping, input.keySpan, ts.DiagnosticCategory.Error,
```

This will limit the diagnostic to highlight the left side of the binding (`[(foo)]`) instead of the whole expression, which looks cleaner.

alxhub reviewed (changes requested):
Last round of comments - after this it should be good to go!

Also, even though this is an unlikely error, it's still a new error and thus we should only land it in a `minor` version instead of a `patch`.

danieltre23 commented:
Flaky failures in presubmit: http://test/OCL:382577037:BASE:383866895:1625849540387:7301e2db forced green.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/diagnostics/src/diagnostic.ts:
```diff
@@ -33,21 +33,26 @@ export interface TemplateDiagnostic extends ts.Diagnostic {
-      }];
+    if (relatedMessages !== undefined) {
+      relatedInformation = [];
+      relatedMessages.forEach(relatedMessage => {
```
```suggestion
      for (const relatedMessage of relatedMessages) {
```

Besides being easier to debug, this will also allow TypeScript to better understand how the types are narrowed here, so you shouldn't need to use `?.push()`.

jessicajaniuk replied:
You could also consider using `.reduce` here since you're transforming the data while iterating through it. That would look like
```
let relatedInformation = relatedMessages.reduce((infoArr, relatedMessage) => {
       infoArr.push({
          category: ts.DiagnosticCategory.Message,
          code: 0,
          file: relatedMessage.sourceFile,
          start: relatedMessage.start,
          length: relatedMessage.end - relatedMessage.start,
          messageText: relatedMessage.text,
       });
       return infoArr;
}, []);
```

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/diagnostics/src/diagnostic.ts:
```diff
@@ -33,21 +33,26 @@ export interface TemplateDiagnostic extends ts.Diagnostic {
+    if (relatedMessages !== undefined) {
+      relatedInformation = [];
+      relatedMessages.forEach(relatedMessage => {
+        relatedInformation?.push({
```
Just as an FYI, this is a situation where you, as a developer, know that `relatedInformation` can never be `undefined` here - it's initialized above. In this case, using the non-null assertion operator (`!`) is more correct than safe navigation (`?.`). But, switching to an actual loop instead of `.forEach` (which we almost never use in the compiler these days) should make this unnecessary.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/diagnostics/src/diagnostic.ts:
```diff
@@ -82,14 +87,16 @@ export function makeTemplateDiagnostic(
-        length: relatedMessage.span.end.offset - relatedMessage.span.start.offset,
-        messageText: relatedMessage.text,
+    if (relatedMessages !== undefined) {
+      relatedMessages.forEach(relatedMessage => {
```
Same here.

alxhub commented:
Reviewed-for: public-api

atscott reviewed (approved):
reviewed-for: public-api

jessicajaniuk commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -1979,6 +1992,31 @@ function tcbEventHandlerExpression(ast: AST, tcb: Context, scope: Scope): ts.Exp
   return translator.translate(ast);
 }
 
+function checkIsSplitTwoWayBinding(
```
The naming here is a bit redundant. `checkIsSplitTwoWayBinding` vs just `isSplitTwoWayBinding`. Both mean the same thing.

jessicajaniuk reviewed (approved):
LGTM :cookie:

I left a couple of suggestions, but nothing I feel strongly about.

reviewed-for: public-api

danieltre23 commented:
@jessicajaniuk updated relatedMessages to use for-of per a separate suggestion from Alex.

petebacondarwin reviewed (approved):
Reviewed-for: public-api

dgp1130 commented:
One final TGP looks good: http://test/OCL:384348576:BASE:384348514:1626141913210:709ae796

zarend commented:
👍 🍌 📦💥
